File: haskell/src/Solutions/Euler103.hs
module Solutions.Euler103 where

twoSubsets' :: [a] -> [([a], [a])]
twoSubsets' [] = [([], [])]
twoSubsets' (h:t) =
    let l = twoSubsets' t
    in l ++ [(h:l1, l2) | (l1, l2) <- l] ++ [(l1, h:l2) | (l1, l2) <- l] 
twoSubsets :: [a] -> [([a], [a])]
twoSubsets = filter (\(x, y) -> not (null x) && not (null y)) . twoSubsets'

p :: [Int] -> [Int] -> Bool
p l1 l2 = case compare (length l1) (length l2) of
    LT -> sum l1 < sum l2
    EQ -> sum l1 /= sum l2
    GT -> sum l1 > sum l2

testSet :: [Int] -> Bool
testSet l = and [p b c | (b, c) <- twoSubsets l]

bound = 50
min' (a, b) (c, d) = if a < c then (a, b) else (c, d)

compute = snd $ foldr1 min' [(sum l, l) |
    a <- [12..bound],
    b <- [a + 1..bound],
    c <- [b + 1..bound],
    d <- [c + 1..bound],
    e <- [d + 1..bound],
    f <- [e + 1..bound],
    g <- [f + 1..bound],
    let l = [a, b, c, d, e, f, g],
    testSet l]

euler103 :: IO String
euler103 = return $ concatMap show compute

File: rust/src/solutions/euler103.rs
use itertools::Itertools;
use std::cmp::Ordering;

fn two_subsets<'a>(set: &'a [u32]) -> Box<dyn Iterator<Item = (Vec<u32>, Vec<u32>)> + 'a> {
    fn f<'a>(set: &'a [u32]) -> Box<dyn Iterator<Item = (Vec<u32>, Vec<u32>)> + 'a> {
        if set.is_empty() {
            Box::new(vec![(vec![], vec![])].into_iter())
        } else {
            Box::new(f(&set[1..]).flat_map(move |(left, right)| {
                let mut lefth = left.clone();
                lefth.push(set[0]);
                let mut righth = right.clone();
                righth.push(set[0]);
                vec![
                    (left.clone(), right.clone()),
                    (lefth, right),
                    (left, righth),
                ]
            }))
        }
    }
    Box::new(f(set).filter(|(left, right)| !left.is_empty() && !right.is_empty()))
}

pub fn test_set(set: &[u32]) -> bool {
    two_subsets(set).all(|(left, right)| {
        let cmp = left.len().cmp(&right.len());
        let left_sum = left.into_iter().sum::<u32>();
        let right_sum = right.into_iter().sum::<u32>();
        match cmp {
            Ordering::Less => left_sum < right_sum,
            Ordering::Equal => left_sum != right_sum,
            Ordering::Greater => left_sum > right_sum,
        }
    })
}

pub fn euler103() -> String {
    (20u32..46)
        .combinations(7)
        .filter(|set| test_set(&set[..]))
        .min_by_key(|v| v.iter().sum::<u32>())
        .unwrap()
        .into_iter()
        .map(|el| el.to_string())
        .collect()
}
